python/pyarrow/src/arrow/python/common.cc

```diff
@@ -131,6 +138,34 @@ class PythonErrorDetail : public StatusDetail {
   }
 
  protected:
+  Result<std::string> FormatImpl() const {
+    PyAcquireGIL lock;
+
+    // Use traceback.format_exception()
+    OwnedRef traceback_module;
+    RETURN_NOT_OK(internal::ImportModule("traceback", &traceback_module));
+
+    OwnedRef fmt_exception;
+    RETURN_NOT_OK(internal::ImportFromModule(traceback_module.obj(), "format_exception",
+                                             &fmt_exception));
+
+    OwnedRef formatted;
+    formatted.reset(PyObject_CallFunctionObjArgs(fmt_exception.obj(), exc_type_.obj(),
+                                                 exc_value_.obj(), exc_traceback_.obj(),
+                                                 NULL));
+
+    std::stringstream ss;
+    ss << "Python exception: ";
+    Py_ssize_t num_lines = PyList_GET_SIZE(formatted.obj());
```

Comment:
You should probably check that we do have a list here. Or you can simply use `PySequence_Length` and `PySequence_GetItem` (this code is not performance-critical).

Reply:
The [API docs](https://docs.python.org/3/library/traceback.html#traceback.format_exception) claim this should return a list, but it's perhaps possible that they might change to some other sequence. I changed to be generic sequences.

I assume the function calls will do the type checking internally, correct?

Reply:
> I assume the function calls will do the type checking internally, correct?

If you're the concrete `PyList` APIs, they will generally assume that it is a list object. Also, fast access macros such as `PyList_GET_ITEM` don't do any error checking.

Reply:
Ah right. I see the distinction now between the error checking and non-error-checking variants. 👍